app/services/structured_fact_importer.rb

```diff
@@ -1,44 +1,73 @@
     facts = facts.select { |k, v| v.present? }
-    normalize_recurse({}, facts)
+    facts = flatten_composite({}, facts)
+
+    original_keys = facts.keys.to_a
+    original_keys.each do |key|
+      fill_hierarchy(facts, key)
+    end
+
+    facts
   end
 
   # expand {'a' => {'b' => 'c'}} to {'a' => nil, 'a::b' => 'c'}
-  def normalize_recurse(memo, facts, prefix = '')
+  def flatten_composite(memo, facts, prefix = '')
     facts.each do |k, v|
       k = prefix.empty? ? k.to_s : prefix + FactName::SEPARATOR + k.to_s
       if v.is_a?(Hash)
         memo[k] = nil
-        normalize_recurse(memo, v, k)
+        flatten_composite(memo, v, k)
       else
         memo[k] = v.to_s
       end
     end
     memo
   end
 
+  # ensures that parent facts already exist in the hash.
+  # Example: for fact: "a::b::c", it will make sure that "a" and "a::b" exist in
+  # the hash.
+  def fill_hierarchy(facts, child_fact_name)
+    facts[child_fact_name] = nil unless facts.key?(child_fact_name)
+
+    parent_name = parent_fact_name(child_fact_name)
+    fill_hierarchy(facts, parent_name) if parent_name
+  end
+
   def preload_fact_names
     # Also fetch compose values, generating {NAME => [ID, COMPOSE]}, avoiding loading the entire model
     Hash[fact_name_class.where(:type => fact_name_class).reorder('').pluck(:name, :id, :compose).map { |fact| [fact.shift, fact] }]
   end
 
-  def find_or_create_fact_name(name, value = nil)
-    if name.include?(FactName::SEPARATOR)
-      parent_name = /(.*)#{FactName::SEPARATOR}/.match(name)[1]
-      parent_fact = find_or_create_fact_name(parent_name, nil)
-      fact_name = parent_fact.children.where(:name => name, :type => fact_name_class.to_s).first
-    else
-      parent_fact = nil
-      fact_name = fact_name_class.where(:name => name, :type => fact_name_class.to_s).first
-    end
+  def ensure_fact_names
+    super
 
-    if fact_name
-      fact_name.update_attribute(:compose, value.nil?) if value.nil? && !fact_name.compose?
-    else
-      fact_name = fact_name_class.create!(:name => name, :type => fact_name_class.to_s, :compose => value.nil?, :parent => parent_fact)
-    end
-    fact_name
+    composite_fact_names = facts.map do |key, value|
+      key if value.nil?
+    end.compact
+
+    affected_records = fact_name_class.where(:name => composite_fact_names, :compose => false).update_all(:compose => true)
+
+    # reload name records if compose flag was reset.
+    initialize_fact_names if affected_records > 0
+  end
+
+  def fact_name_attributes(name)
+    attributes = super
+    fact_value = facts[name]
+    parent_fact_record = fact_names[parent_fact_name(name)]
+
+    attributes.merge(
+      {
+        :parent => parent_fact_record,
+        :compose => fact_value.nil?
+      })
+  end
+
+  def parent_fact_name(child_fact_name)
+    #this regex could be optimized for better performance
+    /(.*)#{FactName::SEPARATOR}/.match(child_fact_name).try(:[], 1)
   end
```

Comment:
For sure put it outside of method so it does not need to be recompiled every run.